lib/puppet/resource.rb

```diff
@@ -431,24 +431,31 @@ def key_attributes
     resource_type.respond_to?(:key_attributes) ? resource_type.key_attributes : [:name]
   end
 
-  # Convert our resource to yaml for Hiera purposes.
-  def to_hierayaml
-    # Collect list of attributes to align => and move ensure first
-    attr = parameters.keys
-    attr_max = attr.inject(0) { |max,k| k.to_s.length > max ? k.to_s.length : max }
-
-    attr.sort!
-    if attr.first != :ensure  && attr.include?(:ensure)
-      attr.delete(:ensure)
-      attr.unshift(:ensure)
+  def hiera_yaml_serialize(data)
+    case data
+    when Array
+      data.map { |v| hiera_yaml_serialize(v) }
+    when Hash
+      data.map do |k, v|
+        { k.to_s => hiera_yaml_serialize(v) }
+      end.inject(:merge)
+    when Time, Symbol
+      data.to_s
+    else
+      data
     end
+  end
 
-    attributes = attr.collect { |k|
-      v = parameters[k]
-      "    %-#{attr_max}s: %s\n" % [k, Puppet::Parameter.format_value_for_display(v)]
-    }.join
+  # Convert our resource to a hash suitable for YAML serialization.
+  def to_hiera_yaml_hash
+    res = {}
+
+    if parameters.include?(:ensure)
+      res['ensure'] = parameters.delete(:ensure).to_s
+    end
+    res.merge!(hiera_yaml_serialize(Hash[parameters.sort]) || {})
```

Comment:
To conform to the puppet style guide,  the `to_hierayaml` returns parameters with `ensure` first and other properties/parameters in the order returned by `parameter.keys`. This looks to try to handle `:ensure` specially, but there's no guarantee about the order of hash keys once `res['ensure']` is called.

Reply:
According to the Ruby [Hash](https://ruby-doc.org/core-2.5.0/Hash.html) documentation:
> Hashes enumerate their values in the order that the corresponding keys were inserted.

So I believe that this handle the `ensure` parameter as expected (an empty hash is created, `ensure` is set, then other values are inserted).

Or did I miss your point?

Reply:
ah right, I forgot that ruby 1.9 changed that. 👍